### PR TITLE
Improve warning for `throw new SomeError()`

### LINKED ISSUE #12529: warning: Throwing unmanaged error - please throw owned
### Summary of Problem

When I write code to throw an error, I get a rather perplexing compiler warning:

`warning: Throwing unmanaged error - please throw owned`

I understand Chapel's default memory ownership to be `borrowed`, and so I'm not sure why the compiler infers it to be unmanaged in this case.

@mppf Brad said you might like to tackle this one.

### Steps to Reproduce

**Source Code:**

```chapel
proc foo() throws {
  throw new IllegalArgumentError("Error message.");
}

foo();
```

**Compile command:**

`chpl foo.chpl`

**Execution command:**

`./foo`

### Configuration Information

- Output of `chpl --version`:

chpl version 1.19.0 pre-release (eecc5d445b)

- Output of `$CHPL_HOME/util/printchplenv --anonymize`:

CHPL_TARGET_PLATFORM: darwin
CHPL_TARGET_COMPILER: clang
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_AUX_FILESYS: none

- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:

Apple LLVM version 10.0.0 (clang-1000.10.44.4)
Target: x86_64-apple-darwin18.2.0
Thread model: posix

### PR DESCRIPTION
**Pull request by mppf**
Resolves #12529.

Improves the error for `throw new SomeError()` from e.g.
```
foo.chpl:3: warning: Throwing unmanaged error - please throw owned
```

to

```
foo.chpl:3: warning: please use 'throw new owned IllegalArgumentError instead of 'throw new IllegalArgumentError'
```

Additionally adds testing for various invalid cases of `throw`.

- [x] full local testing

Reviewed by @lydia-duncan - thanks!

### CODE REVIEW

**Comment by lydia-duncan** [comment]
> It'd be hard to remove the 2nd warning completely (but we could change the text of it).

Hmm.  The way it is phrased or the fact that it is also a warning kinda makes it seem like they are separate problems.  Can we make it a note to avoid confusion, maybe `'throw new Error' throws an unmanaged Error`?

**Comment by dlongnecke-cray** [comment]
> > It'd be hard to remove the 2nd warning completely (but we could change the text of it).
> 
> Hmm. The way it is phrased or the fact that it is also a warning kinda makes it seem like they are separate problems. Can we make it a note to avoid confusion, maybe `'throw new Error' throws an unmanaged Error`?

This would make me happy. As long as I have some sort of explanation for how the compiler got to `unmanaged` when the intent was `borrowed`, I don't feel like I need to know more.

File: compiler/resolution/functionResolution.cpp [hunk 1]
```diff
@@ -5786,6 +5790,7 @@ static void resolveNewSetupManaged(CallExpr* newExpr, Type*& manager) {
         }
       }
 
+
```
Comment (lydia-duncan) [inline compiler/resolution/functionResolution.cpp]: Unnecessary extra line?

File: compiler/resolution/functionResolution.cpp [hunk 2]
```diff
@@ -5898,6 +5898,66 @@ static bool isUndecoratedClassNew(CallExpr* newExpr, Type* newType) {
   return false;
 }
 
+static void warnForThrowNotOwned(CallExpr* newExpr, Type* newType, Type* manager) {
+  // This function implements a warning that should be removed in 1.20
+  // in favor e.g. calling compilerError in chpl_fix_thrown_error.
```
Comment (lydia-duncan) [inline compiler/resolution/functionResolution.cpp]: Missing "of" before/after the "e.g."?

**Comment by mppf** [comment]
I'm testing PR #12533, which now removes the confusing 2nd line (at the cost of more code in the compiler).

File: compiler/resolution/functionResolution.cpp [hunk 3]
```diff
@@ -5898,6 +5898,66 @@ static bool isUndecoratedClassNew(CallExpr* newExpr, Type* newType) {
...
+                       "please use 'throw new owned %s "
+                       "instead of 'throw new %s'",
+                       cType->symbol->name, cType->symbol->name);
+            else if (unmanaged)
+              USR_WARN(parentCall,
+                       "please throw owned %s instead of unmanaged %s",
+                       cType->symbol->name, cType->symbol->name);
+            else if (borrowed)
+              USR_FATAL(parentCall,
+                       "please throw owned %s instead borrowed %s",
```
Comment (lydia-duncan) [inline compiler/resolution/functionResolution.cpp]: "instead of borrowed"

Comment (lydia-duncan) [reply compiler/resolution/functionResolution.cpp]: Also, there's inconsistency here in the use of '.  The first message is missing a terminating ' for the owned section, and the second and third don't use ' at all.  I'd prefer they all use '.

**Comment by mppf** [comment]
thanks @lydia-duncan - FYI I plan to merge it once the next testing run completes.

**Pull request merged**